File: src/health/dns.rs
//! DNS health probe socket.
//!
//! Sends minimal DNS queries (root "." A record) to configured DNS server
//! targets over UDP port 53. Any valid DNS response with a matching
//! transaction ID indicates the server is reachable.
//!
//! Like ICMP probes, DNS sockets are bound to a specific device via
//! `SO_BINDTODEVICE` and marked with `SO_MARK = 0xDEAD` for nftables
//! exemption.

use std::io;
use std::mem;
use std::net::IpAddr;
use std::os::unix::io::RawFd;

use crate::error::{Error, Result};
use super::ProbeTransport;

/// Firewall mark applied to all health-probe packets (shared with ICMP).
const PROBE_MARK: u32 = 0xDEAD;

/// Standard DNS port.
const DNS_PORT: u16 = 53;

/// UDP socket for DNS health probes.
pub struct DnsSocket {
    fd: RawFd,
    /// Interface identifier returned on recv to match the probe engine's
    /// (seq, id) protocol. Stored at construction time since DNS responses
    /// don't carry this information.
    probe_id: u16,
    /// Pre-built DNS query template (txid=0). Bytes 0-1 are overwritten per send.
    query_template: Vec<u8>,
}

impl DnsSocket {
    /// Create a UDP socket for DNS probes over IPv4.
    pub fn new_v4(device: &str, probe_id: u16, query_name: &str) -> Result<Self> {
        let fd = unsafe {
            libc::socket(
                libc::AF_INET,
                libc::SOCK_DGRAM | libc::SOCK_CLOEXEC,
                0,
            )
        };
        if fd < 0 {
            return Err(Error::Io(io::Error::last_os_error()));
        }

        if let Err(e) = Self::configure(fd, device) {
            unsafe { libc::close(fd) };
            return Err(e);
        }

        let query_template = build_dns_query_for(0, query_name);
        Ok(Self { fd, probe_id, query_template })
    }

    /// Create a UDP socket for DNS probes over IPv6.
    pub fn new_v6(device: &str, probe_id: u16, query_name: &str) -> Result<Self> {
        let fd = unsafe {
            libc::socket(
                libc::AF_INET6,
                libc::SOCK_DGRAM | libc::SOCK_CLOEXEC,
                0,
            )
        };
        if fd < 0 {
            return Err(Error::Io(io::Error::last_os_error()));
        }

        if let Err(e) = Self::configure(fd, device) {
            unsafe { libc::close(fd) };
            return Err(e);
        }

        let query_template = build_dns_query_for(0, query_name);
        Ok(Self { fd, probe_id, query_template })
    }

    /// Apply common socket options: SO_MARK, SO_BINDTODEVICE, non-blocking.
    fn configure(fd: RawFd, device: &str) -> Result<()> {
        let mark = PROBE_MARK;
        let ret = unsafe {
            libc::setsockopt(
                fd,
                libc::SOL_SOCKET,
                libc::SO_MARK,
                &mark as *const u32 as *const libc::c_void,
                mem::size_of::<u32>() as libc::socklen_t,
            )
        };
        if ret < 0 {
            return Err(Error::Io(io::Error::last_os_error()));
        }

        let dev_bytes = device.as_bytes();
        if dev_bytes.len() >= libc::IFNAMSIZ {
            return Err(Error::Config(format!(
                "device name too long: {device}"
            )));
        }
        let ret = unsafe {
            libc::setsockopt(
                fd,
                libc::SOL_SOCKET,
                libc::SO_BINDTODEVICE,
                dev_bytes.as_ptr() as *const libc::c_void,
                dev_bytes.len() as libc::socklen_t,
            )
        };
        if ret < 0 {
            return Err(Error::Io(io::Error::last_os_error()));
        }

        let flags = unsafe { libc::fcntl(fd, libc::F_GETFL) };
        if flags < 0 {
            return Err(Error::Io(io::Error::last_os_error()));
        }
        let ret = unsafe { libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK) };
        if ret < 0 {
            return Err(Error::Io(io::Error::last_os_error()));
        }

        Ok(())
    }

    /// Send a DNS query to `target:53`.
    ///
    /// Builds a minimal query for "." (root) type A. The DNS transaction ID
    /// is set to `seq` so the response can be matched back to the probe.
    pub fn send_query(&mut self, target: IpAddr, seq: u16, _id: u16) -> Result<()> {
        // Stamp the txid into the cached template (bytes 0-1)
        self.query_template[0] = (seq >> 8) as u8;
        self.query_template[1] = (seq & 0xFF) as u8;
        let query = &self.query_template;

        match target {
            IpAddr::V4(addr) => {
                let octets = addr.octets();
                let mut sa: libc::sockaddr_in = unsafe { mem::zeroed() };
                sa.sin_family = libc::AF_INET as libc::sa_family_t;
                sa.sin_port = DNS_PORT.to_be();
                sa.sin_addr.s_addr = u32::from_ne_bytes(octets);

                let ret = unsafe {
                    libc::sendto(
                        self.fd,
                        query.as_ptr() as *const libc::c_void,
                        query.len(),
                        0,
                        &sa as *const libc::sockaddr_in as *const libc::sockaddr,
                        mem::size_of::<libc::sockaddr_in>() as libc::socklen_t,
                    )
                };
                if ret < 0 {
                    return Err(Error::Io(io::Error::last_os_error()));
                }
            }
            IpAddr::V6(addr) => {
                let mut sa: libc::sockaddr_in6 = unsafe { mem::zeroed() };
                sa.sin6_family = libc::AF_INET6 as libc::sa_family_t;
                sa.sin6_port = DNS_PORT.to_be();
                sa.sin6_addr.s6_addr = addr.octets();

                let ret = unsafe {
                    libc::sendto(
                        self.fd,
                        query.as_ptr() as *const libc::c_void,
                        query.len(),
                        0,
                        &sa as *const libc::sockaddr_in6 as *const libc::sockaddr,
                        mem::size_of::<libc::sockaddr_in6>() as libc::socklen_t,
                    )
                };
                if ret < 0 {
                    return Err(Error::Io(io::Error::last_os_error()));
                }
            }
        }

        Ok(())
    }

    /// Non-blocking receive of a DNS response.
    ///
    /// Returns `Ok(Some((seq, id)))` if a valid DNS response was received
    /// (transaction ID maps to seq, id is the stored probe_id),
    /// `Ok(None)` if no data is available (EAGAIN), or an error.
    pub fn recv_response(&self) -> Result<Option<(u16, u16)>> {
        let mut buf = [0u8; 512];
        let ret = unsafe {
            libc::recv(
                self.fd,
                buf.as_mut_ptr() as *mut libc::c_void,
                buf.len(),
                0,
            )
        };
        if ret < 0 {
            let err = io::Error::last_os_error();
            if err.kind() == io::ErrorKind::WouldBlock {
                return Ok(None);
            }
            return Err(Error::Io(err));
        }

        let n = ret as usize;

        // DNS header is 12 bytes minimum
        if n < 12 {
            return Ok(None);
        }

        // Check QR bit (bit 15 of flags) -- must be 1 (response)
        if buf[2] & 0x80 == 0 {
            return Ok(None);
        }

        // Extract transaction ID
        let txid = u16::from_be_bytes([buf[0], buf[1]]);

        Ok(Some((txid, self.probe_id)))
    }

    /// Return the raw file descriptor for mio registration.
    pub fn fd(&self) -> RawFd {
        self.fd
    }
}

impl ProbeTransport for DnsSocket {
    fn send(&mut self, target: IpAddr, seq: u16, id: u16, _payload_size: usize) -> Result<()> {
        self.send_query(target, seq, id)
    }
    fn recv(&mut self) -> Result<Option<(u16, u16)>> {
        self.recv_response()
    }
    fn fds(&self) -> Vec<RawFd> {
        vec![self.fd()]
    }
}

impl Drop for DnsSocket {
    fn drop(&mut self) {
        unsafe { libc::close(self.fd) };
    }
}

/// Build a DNS query for a given domain name (or root "." if empty).
///
/// Layout:
///   - Header (12 bytes): txid, flags=0x0100 (standard query, RD=1), qdcount=1
///   - Question: encoded domain name, type A (1), class IN (1)
fn build_dns_query_for(txid: u16, name: &str) -> Vec<u8> {
    let mut q = vec![0u8; 12];

    // Transaction ID
    q[0] = (txid >> 8) as u8;
    q[1] = (txid & 0xFF) as u8;

    // Flags: standard query, recursion desired
    q[2] = 0x01;
    q[3] = 0x00;

    // QDCOUNT = 1
    q[4] = 0x00;
    q[5] = 0x01;

    // Question section: encode domain name as DNS labels
    if name.is_empty() {
        q.push(0x00); // root label
    } else {
        for label in name.trim_end_matches('.').split('.') {
            let len = label.len().min(63) as u8;
            q.push(len);
            q.extend_from_slice(&label.as_bytes()[..len as usize]);
        }
        q.push(0x00); // terminating root label
    }

    // Type A (1)
    q.push(0x00);
    q.push(0x01);

    // Class IN (1)
    q.push(0x00);
    q.push(0x01);

    q
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dns_query_structure() {
        let q = build_dns_query_for(0x1234, "");

        // Transaction ID
        assert_eq!(q[0], 0x12);
        assert_eq!(q[1], 0x34);

        // Flags: standard query, RD=1
        assert_eq!(q[2], 0x01);
        assert_eq!(q[3], 0x00);

        // QDCOUNT = 1
        assert_eq!(u16::from_be_bytes([q[4], q[5]]), 1);

        // ANCOUNT, NSCOUNT, ARCOUNT = 0
        assert_eq!(u16::from_be_bytes([q[6], q[7]]), 0);
        assert_eq!(u16::from_be_bytes([q[8], q[9]]), 0);
        assert_eq!(u16::from_be_bytes([q[10], q[11]]), 0);

        // Root label
        assert_eq!(q[12], 0x00);

        // Type A
        assert_eq!(u16::from_be_bytes([q[13], q[14]]), 1);

        // Class IN
        assert_eq!(u16::from_be_bytes([q[15], q[16]]), 1);

        // Total length
        assert_eq!(q.len(), 17);
    }

    #[test]
    fn dns_query_with_domain_name() {
        let q = build_dns_query_for(0xABCD, "example.com");

        // Transaction ID
        assert_eq!(q[0], 0xAB);
        assert_eq!(q[1], 0xCD);

        // Question section starts at byte 12
        // "example" label: length 7, then "example"
        assert_eq!(q[12], 7);
        assert_eq!(&q[13..20], b"example");

        // "com" label: length 3, then "com"
        assert_eq!(q[20], 3);
        assert_eq!(&q[21..24], b"com");

        // Root terminator
        assert_eq!(q[24], 0x00);

        // Type A
        assert_eq!(u16::from_be_bytes([q[25], q[26]]), 1);

        // Class IN
        assert_eq!(u16::from_be_bytes([q[27], q[28]]), 1);

        // Total: 12 header + 1+7 + 1+3 + 1 + 2+2 = 29
        assert_eq!(q.len(), 29);
    }

    #[test]
    fn dns_query_trailing_dot_stripped() {
        let q1 = build_dns_query_for(1, "example.com.");
        let q2 = build_dns_query_for(1, "example.com");
        // Trailing dot should be stripped, producing identical question sections
        assert_eq!(q1[12..], q2[12..]);
    }
}

File: src/health/dns.nim
## DNS health probe socket.
##
## Sends minimal DNS queries to configured DNS server targets over UDP
## port 53. Any valid DNS response with a matching transaction ID
## indicates the server is reachable.
##
## Like ICMP probes, DNS sockets are bound to a specific device via
## SO_BINDTODEVICE and marked with SO_MARK = 0xDEAD for nftables
## exemption.

import std/[posix, os]
import ../linux_constants

const DNS_PORT = 53'u16

proc encodeDnsQuery*(name: string, buf: var array[512, byte]): int =
  ## Encode a DNS A query. Returns bytes written.
  ##
  ## Layout:
  ##   - Header (12 bytes): txid=0, flags=0x0100 (standard query, RD=1), qdcount=1
  ##   - Question: label-encoded name + type A (0x0001) + class IN (0x0001)
  var pos = 0

  # Transaction ID (0 -- filled at send time)
  buf[0] = 0; buf[1] = 0
  pos = 2

  # Flags: standard query, recursion desired (0x0100)
  buf[2] = 0x01; buf[3] = 0x00
  pos = 4

  # QDCOUNT = 1
  buf[4] = 0x00; buf[5] = 0x01
  pos = 6

  # ANCOUNT, NSCOUNT, ARCOUNT = 0
  buf[6] = 0; buf[7] = 0   # ANCOUNT
  buf[8] = 0; buf[9] = 0   # NSCOUNT
  buf[10] = 0; buf[11] = 0 # ARCOUNT
  pos = 12

  # Question section: encode domain name as DNS labels
  # Trim leading and trailing dots (inline to avoid std/strutils dependency)
  var trimStart = 0
  var trimEnd = name.len - 1
  while trimStart <= trimEnd and name[trimStart] == '.': inc trimStart
  while trimEnd >= trimStart and name[trimEnd] == '.': dec trimEnd
  let trimmed = if trimStart > trimEnd: "" else: name[trimStart .. trimEnd]
  if trimmed.len == 0:
    # Root query: single null byte
    buf[pos] = 0x00
    pos += 1
  else:
    # Split on '.' and encode each label (with bounds checking)
    var start = 0
    while start <= trimmed.len:
      var dotPos = trimmed.len
      for i in start ..< trimmed.len:
        if trimmed[i] == '.':
          dotPos = i
          break
      let labelLen = dotPos - start
      let writeLen = min(labelLen, 63)
      # Bounds check: need 1 (length byte) + writeLen + 1 (null term) + 4 (type+class)
      if pos + 1 + writeLen + 1 + 4 > buf.len:
        break  # truncate — name too long for buffer
      buf[pos] = uint8(writeLen)
      pos += 1
      for i in 0 ..< writeLen:
        buf[pos] = uint8(trimmed[start + i])
        pos += 1
      start = dotPos + 1
    # Terminating root label
    if pos < buf.len:
      buf[pos] = 0x00
      pos += 1

  # Type A (0x0001) + Class IN (0x0001)
  if pos + 4 <= buf.len:
    buf[pos] = 0x00; buf[pos + 1] = 0x01
    pos += 2
    buf[pos] = 0x00; buf[pos + 1] = 0x01
    pos += 2

  result = pos

proc createDnsSocket*(device: string, family: uint8): cint =
  ## Create a UDP socket for DNS probes.
  ## `family` is AF_INET or AF_INET6. Returns raw fd.
  let af = if family == AF_INET.uint8: AF_INET.cint else: AF_INET6.cint

  result = cint(socket(af, SOCK_DGRAM.cint, 0))
  if result < 0:
    raiseOSError(osLastError())

  let fd = result

  # SO_MARK
  var mark = PROBE_MARK
  if setsockopt(SocketHandle(fd), SOL_SOCKET.cint, linux_constants.SO_MARK.cint,
                addr mark, sizeof(mark).SockLen) < 0:
    discard close(fd)
    raiseOSError(osLastError())

  # SO_BINDTODEVICE
  if setsockopt(SocketHandle(fd), SOL_SOCKET.cint, linux_constants.SO_BINDTODEVICE.cint,
                cstring(device), device.len.SockLen) < 0:
    discard close(fd)
    raiseOSError(osLastError())

  # Set non-blocking
  let flags = fcntl(fd, F_GETFL)
  if flags < 0:
    discard close(fd)
    raiseOSError(osLastError())
  if fcntl(fd, F_SETFL, flags or O_NONBLOCK) < 0:
    discard close(fd)
    raiseOSError(osLastError())

proc sendDnsProbe*(fd: cint, target: openArray[byte], family: uint8,
                   txid: uint16, queryBuf: var openArray[byte],
                   queryLen: int): bool =
  ## Stamp txid into bytes 0-1 of query, sendto() to port 53.
  ## Returns true on success.
  queryBuf[0] = uint8(txid shr 8)
  queryBuf[1] = uint8(txid and 0xFF)

  if family == AF_INET.uint8:
    var sa: Sockaddr_in
    sa.sin_family = AF_INET.TSa_Family
    sa.sin_port = htons(DNS_PORT)
    copyMem(addr sa.sin_addr, unsafeAddr target[0], 4)
    let ret = sendto(SocketHandle(fd), cast[pointer](addr queryBuf[0]), queryLen, 0'i32,
                     cast[ptr SockAddr](addr sa),
                     sizeof(Sockaddr_in).SockLen)
    result = ret >= 0
  else:
    var sa: Sockaddr_in6
    sa.sin6_family = AF_INET6.TSa_Family
    sa.sin6_port = htons(DNS_PORT)
    copyMem(addr sa.sin6_addr, unsafeAddr target[0], 16)
    let ret = sendto(SocketHandle(fd), cast[pointer](addr queryBuf[0]), queryLen, 0'i32,
                     cast[ptr SockAddr](addr sa),
                     sizeof(Sockaddr_in6).SockLen)
    result = ret >= 0

proc recvDnsReply*(fd: cint, buf: var array[512, byte]): tuple[ok: bool, txid: uint16] =
  ## Non-blocking receive of a DNS response.
  ## Checks QR bit (byte 2, bit 7), extracts txid from bytes 0-1.
  result = (ok: false, txid: 0'u16)

  let n = recv(SocketHandle(fd), cast[pointer](addr buf[0]), buf.len, 0)
  if n < 12:
    return  # DNS header is 12 bytes minimum

  # Check QR bit (bit 7 of byte 2) -- must be 1 (response)
  if (buf[2] and 0x80) == 0:
    return

  let txid = (uint16(buf[0]) shl 8) or uint16(buf[1])
  result = (ok: true, txid: txid)

when isMainModule:
  # Tests for encodeDnsQuery

  block testGoogleCom:
    ## "google.com" produces correct label sequence.
    var buf: array[512, byte]
    let n = encodeDnsQuery("google.com", buf)

    # Header: txid=0, flags=0x0100, qdcount=1
    doAssert buf[0] == 0 and buf[1] == 0, "txid should be 0"
    doAssert buf[2] == 0x01 and buf[3] == 0x00, "flags should be 0x0100"
    doAssert buf[4] == 0x00 and buf[5] == 0x01, "qdcount should be 1"

    # Question starts at byte 12
    # "google" label: length 6
    doAssert buf[12] == 6, "google label length"
    doAssert buf[13] == uint8('g')
    doAssert buf[14] == uint8('o')
    doAssert buf[15] == uint8('o')
    doAssert buf[16] == uint8('g')
    doAssert buf[17] == uint8('l')
    doAssert buf[18] == uint8('e')

    # "com" label: length 3
    doAssert buf[19] == 3, "com label length"
    doAssert buf[20] == uint8('c')
    doAssert buf[21] == uint8('o')
    doAssert buf[22] == uint8('m')

    # Root terminator
    doAssert buf[23] == 0x00, "root terminator"

    # Type A
    doAssert buf[24] == 0x00 and buf[25] == 0x01, "type A"

    # Class IN
    doAssert buf[26] == 0x00 and buf[27] == 0x01, "class IN"

    # Total: 12 header + 1+6 + 1+3 + 1 + 2+2 = 28
    doAssert n == 28, "total length for google.com should be 28"

  block testRootQuery:
    ## Root query "." produces single null byte after header.
    var buf: array[512, byte]
    let n = encodeDnsQuery(".", buf)

    # Root label (single null byte)
    doAssert buf[12] == 0x00, "root label should be null"

    # Type A
    doAssert buf[13] == 0x00 and buf[14] == 0x01, "type A"

    # Class IN
    doAssert buf[15] == 0x00 and buf[16] == 0x01, "class IN"

    # Total: 12 + 1 + 2 + 2 = 17
    doAssert n == 17, "total length for root query should be 17"

  block testTrailingDot:
    ## Name with trailing dot "example.com." works correctly.
    var buf1: array[512, byte]
    var buf2: array[512, byte]
    let n1 = encodeDnsQuery("example.com.", buf1)
    let n2 = encodeDnsQuery("example.com", buf2)

    # Trailing dot should be stripped, producing identical output
    doAssert n1 == n2, "trailing dot should not change length"
    for i in 12 ..< n1:
      doAssert buf1[i] == buf2[i],
        fmt"trailing dot: byte {i} differs"

  echo "All encodeDnsQuery tests passed."
